File: EVTs/adder/libadder-java/src/edu/uconn/cse/adder/VoteProof.java
package edu.uconn.cse.adder;

import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class VoteProof {
    private AdderInteger p;
    private List/*<MembershipProof>*/ proofList;
    private MembershipProof sumProof;

    public VoteProof() {

    }

    private VoteProof(MembershipProof sumProof,
                      List/*<MembershipProof>*/ proofList) {
        this.sumProof = sumProof;
        this.proofList = proofList;
    }

    public void compute(Vote vote, PublicKey pubKey,
                        List/*<AdderInteger>*/ choices, int min, int max) {
        this.p = pubKey.getP();
        List/*<ElgamalCiphertext>*/ cipherList = vote.getCipherList();
        List/*<AdderInteger>*/ cipherDomain
            = new ArrayList/*<AdderInteger>*/(2);

        cipherDomain.add(AdderInteger.ZERO);
        cipherDomain.add(AdderInteger.ONE);

        ElgamalCiphertext sumCipher
            = new ElgamalCiphertext(AdderInteger.ONE, AdderInteger.ONE, p);

        int numChoices = 0;
        int size = cipherList.size();
        this.proofList = new ArrayList(size);

        for (int i = 0; i < size; i++) {
            MembershipProof proof = new MembershipProof();
            ElgamalCiphertext ciphertext
                = (ElgamalCiphertext) cipherList.get(i);
            AdderInteger choice = (AdderInteger) choices.get(i);
            proof.compute(ciphertext, pubKey, choice, cipherDomain);
            this.proofList.add(proof);

            sumCipher = sumCipher.multiply(ciphertext);

            if (choice.equals(AdderInteger.ONE)) {
                numChoices++;
            }
        }

        List/*<AdderInteger>*/ totalDomain
            = new ArrayList/*<AdderInteger>*/(max + 1);

        for (int j = min; j <= max; j++) {
            totalDomain.add(new AdderInteger(j));
        }

        this.sumProof = new MembershipProof();
        this.sumProof.compute(sumCipher, pubKey, new AdderInteger(numChoices), totalDomain);
    }


    public boolean verify(Vote vote, PublicKey pubKey, int min, int max) {
        this.p = pubKey.getP();
        List/*<ElgamalCiphertext>*/ cipherList = vote.getCipherList();
        List/*<AdderInteger>*/ cipherDomain
            = new ArrayList/*<AdderInteger>*/(2);

        cipherDomain.add(AdderInteger.ZERO);
        cipherDomain.add(AdderInteger.ONE);

        ElgamalCiphertext sumCipher
            = new ElgamalCiphertext(AdderInteger.ONE, AdderInteger.ONE, p);
        int size = this.proofList.size();

        for (int i = 0; i < size; i++) {
            MembershipProof proof = (MembershipProof) this.proofList.get(i);
            ElgamalCiphertext ciphertext
                = (ElgamalCiphertext) cipherList.get(i);

            if (!proof.verify(ciphertext, pubKey, cipherDomain)) {
                return false;
            }

            sumCipher = sumCipher.multiply(ciphertext);
        }

        List/*<AdderInteger>*/ totalDomain
            = new ArrayList/*<AdderInteger>*/(max + 1);

        for (int j = min; j <= max; j++) {
            totalDomain.add(new AdderInteger(j));
        }

        if (!this.sumProof.verify(sumCipher, pubKey, totalDomain)) {
            return false;
        }

        return true;
    }

    public static VoteProof fromString(String s) {
        StringTokenizer st = new StringTokenizer(s, " ");
        List/*<ElgamalCiphertext>*/ pList
            = new ArrayList/*<ElgamalCiphertext>*/(25); // XXX: what size? 
        MembershipProof sumProof = MembershipProof.fromString(st.nextToken());

        while (st.hasMoreTokens()) {
            String s2 = st.nextToken();
            MembershipProof proof = MembershipProof.fromString(s2);
            pList.add(proof);
        }

        VoteProof voteProof = new VoteProof(sumProof, pList);

        return voteProof;
    }

    public String toString() {
        StringBuffer sb = new StringBuffer(4096);

        sb.append(sumProof.toString());

        for (int i = 0; i < proofList.size(); i++) {
            MembershipProof proof = (MembershipProof) proofList.get(i);
            sb.append(" ");
            sb.append(proof.toString());
        }

        return sb.toString();
    }
}

File: EVTs/adder/libadder/src/VoteProof.cpp
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*  libadder - Cryptographic library for the Adder project.
    Copyright (C) 2004-2006  The Adder Team <adder@cse.uconn.edu>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/**
 * @file VoteProof.cpp
 * Implementation of the VoteProof class.
 */

#include <cstdlib>
#include <iostream>
#include <vector>

#include <boost/tokenizer.hpp>

#include "common.h"
#include "PublicKey.h"
#include "PrivateKey.h"
#include "VoteProof.h"
#include "Integer.h"
#include "radix64.h"
#include "misc.h"
#include "exceptions.h"

using namespace std;
using namespace boost;

namespace adder 
{
    VoteProof::VoteProof()
    {
        p = Integer("0", 10);
    }

    void VoteProof::compute(Vote vote, 
                            PublicKey pub_key,
                            vector<bool> choices,
                            Integer minimum,
                            Integer maximum)
    {
        p = pub_key.get_p();
        vector<ElgamalCiphertext> cipher_vec = vote.get_cipher_vec();

        vector<Integer> cipher_domain;
        cipher_domain.push_back(0);
        cipher_domain.push_back(1);

        ElgamalCiphertext sum_cipher(1, 1, p);
        Integer num_choices = 0;

        for (unsigned int i = 0; i < cipher_vec.size(); i++) {
            MembershipProof proof;
            proof.compute(cipher_vec[i], pub_key,
                          choices[i] ? 1 : 0, cipher_domain);
            proof_vec.push_back(proof);

            sum_cipher = sum_cipher * cipher_vec[i];

            num_choices = num_choices + (choices[i] ? 1 : 0);
        }

        vector<Integer> total_domain;
        for (Integer j = minimum; j <= maximum; j = j + 1) {
            total_domain.push_back(j);
        }

        MembershipProof proof;
        proof.compute(sum_cipher, pub_key,
                      num_choices, total_domain);

        sum_proof = proof;
    }

    bool VoteProof::verify(Vote vote,
                           PublicKey pub_key,
                           Integer minimum,
                           Integer maximum)
    {
        p = pub_key.get_p();

        vector<ElgamalCiphertext> cipher_vec = vote.get_cipher_vec();

        vector<Integer> cipher_domain;
        cipher_domain.push_back(0);
        cipher_domain.push_back(1);

        ElgamalCiphertext sum_cipher(1, 1, p);

        for (unsigned int i = 0; i < proof_vec.size(); i++) {
            if (!proof_vec[i].verify(cipher_vec[i],
                                     pub_key,
                                     cipher_domain)) {
                return false;
            }

            sum_cipher = sum_cipher * cipher_vec[i];
        }

        vector<Integer> total_domain;
        for (Integer j = minimum; j <= maximum; j = j + 1) {
            total_domain.push_back(j);
        }

        if (!sum_proof.verify(sum_cipher,
                              pub_key,
                              total_domain)) {
            return false;
        }

        return true;
    }

    std::string VoteProof::str() const
    {
        stringstream proof_ss;
        proof_ss << sum_proof.str();

        for (unsigned int i = 0; i < proof_vec.size(); i++) {
            proof_ss << " " << proof_vec[i].str();
        }

        return proof_ss.str();
    }

    void VoteProof::from_str(std::string proof_string)
    {
        tokenizer<> tok(proof_string);
        tokenizer<>::iterator t_iter;

        if (tok.begin() == tok.end()) {
            throw Invalid_proof("empty string");
        }

        t_iter = tok.begin();
        sum_proof.from_str(*t_iter);
        t_iter++;

        for (; t_iter != tok.end(); t_iter++) {
            MembershipProof proof;
            proof.from_str(*t_iter);
            proof_vec.push_back(proof);
        }
    }
}
